File: tests/tenant.test.py
#!/usr/bin/env python3
"""Unit tests for src/tenant.py — mirrors tests/tenant.test.ts shape.

Run: python3 tests/tenant.test.py
"""
import os
import sys
import socket
import tempfile
import shutil
import unittest
from pathlib import Path

# Make `src/` importable as a flat package (mirrors how scripts under src/
# already do `from util_paths import ...`).
REPO_ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO_ROOT / "src"))

from tenant import tenant_id, tenant_path, _reset_for_tests  # type: ignore  # noqa: E402

HOST = socket.gethostname().split(".")[0]


class TenantIdResolution(unittest.TestCase):
    def setUp(self) -> None:
        self.saved = {k: os.environ.get(k) for k in ("SUTANDO_TENANT_ID", "SUTANDO_PRIVATE_DIR", "SUTANDO_WORKSPACE")}
        _reset_for_tests()
        self.tmp = tempfile.mkdtemp(prefix="sutando-tenant-py-")
        os.environ.pop("SUTANDO_TENANT_ID", None)
        os.environ.pop("SUTANDO_PRIVATE_DIR", None)
        os.environ["SUTANDO_WORKSPACE"] = self.tmp

    def tearDown(self) -> None:
        for k, v in self.saved.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v
        _reset_for_tests()
        shutil.rmtree(self.tmp, ignore_errors=True)

    def test_default_when_no_env_no_state(self) -> None:
        self.assertEqual(tenant_id(), "default")

    def test_env_var_when_valid(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "acme"
        self.assertEqual(tenant_id(), "acme")

    def test_invalid_env_falls_back(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "BAD ID!"
        self.assertEqual(tenant_id(), "default")

    def test_reads_state_file(self) -> None:
        state_dir = Path(self.tmp) / "state"
        state_dir.mkdir(parents=True, exist_ok=True)
        (state_dir / "tenant.json").write_text('{"id":"beta-tenant"}')
        self.assertEqual(tenant_id(), "beta-tenant")

    def test_env_wins_over_state(self) -> None:
        state_dir = Path(self.tmp) / "state"
        state_dir.mkdir(parents=True, exist_ok=True)
        (state_dir / "tenant.json").write_text('{"id":"from-file"}')
        os.environ["SUTANDO_TENANT_ID"] = "from-env"
        self.assertEqual(tenant_id(), "from-env")

    def test_corrupt_state_falls_back(self) -> None:
        state_dir = Path(self.tmp) / "state"
        state_dir.mkdir(parents=True, exist_ok=True)
        (state_dir / "tenant.json").write_text("not json {{")
        self.assertEqual(tenant_id(), "default")

    def test_cached_after_first_call(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "first"
        self.assertEqual(tenant_id(), "first")
        os.environ["SUTANDO_TENANT_ID"] = "second"
        self.assertEqual(tenant_id(), "first")
        _reset_for_tests()
        self.assertEqual(tenant_id(), "second")


class TenantPathResolution(unittest.TestCase):
    def setUp(self) -> None:
        self.saved = {k: os.environ.get(k) for k in ("SUTANDO_TENANT_ID", "SUTANDO_PRIVATE_DIR", "SUTANDO_WORKSPACE")}
        _reset_for_tests()
        self.priv = tempfile.mkdtemp(prefix="sutando-priv-py-")
        self.ws = tempfile.mkdtemp(prefix="sutando-ws-py-")
        os.environ["SUTANDO_PRIVATE_DIR"] = self.priv
        os.environ["SUTANDO_WORKSPACE"] = self.ws
        os.environ.pop("SUTANDO_TENANT_ID", None)

    def tearDown(self) -> None:
        for k, v in self.saved.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v
        _reset_for_tests()
        shutil.rmtree(self.priv, ignore_errors=True)
        shutil.rmtree(self.ws, ignore_errors=True)

    def test_shared_under_tenant_when_present(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "acme"
        _reset_for_tests()
        tenant_dir = Path(self.priv) / "tenant-acme"
        tenant_dir.mkdir(parents=True)
        (tenant_dir / "build_log.md").write_text("acme log")
        p = tenant_path("build_log.md", "shared")
        self.assertEqual(p, tenant_dir / "build_log.md")

    def test_device_under_tenant_machine_when_present(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "acme"
        _reset_for_tests()
        machine_dir = Path(self.priv) / "tenant-acme" / f"machine-{HOST}"
        machine_dir.mkdir(parents=True)
        (machine_dir / "pending-questions.md").write_text("q")
        p = tenant_path("pending-questions.md", "device")
        self.assertEqual(p, machine_dir / "pending-questions.md")

    def test_legacy_fallback_shared(self) -> None:
        (Path(self.priv) / "notes.md").write_text("legacy notes")
        p = tenant_path("notes.md", "shared")
        self.assertEqual(p, Path(self.priv) / "notes.md")

    def test_legacy_fallback_device(self) -> None:
        legacy = Path(self.priv) / f"machine-{HOST}"
        legacy.mkdir()
        (legacy / "stand-identity.json").write_text("{}")
        p = tenant_path("stand-identity.json", "device")
        self.assertEqual(p, legacy / "stand-identity.json")

    def test_workspace_fallback(self) -> None:
        (Path(self.ws) / "README.md").write_text("in workspace")
        p = tenant_path("README.md", "shared")
        self.assertEqual(p, Path(self.ws) / "README.md")

    def test_preferred_private_path_on_miss(self) -> None:
        os.environ["SUTANDO_TENANT_ID"] = "acme"
        _reset_for_tests()
        p = tenant_path("does-not-exist.md", "shared")
        self.assertEqual(p, Path(self.priv) / "tenant-acme" / "does-not-exist.md")

    def test_no_private_dir_returns_workspace(self) -> None:
        os.environ.pop("SUTANDO_PRIVATE_DIR", None)
        (Path(self.ws) / "x.md").write_text("x")
        self.assertEqual(tenant_path("x.md", "shared"), Path(self.ws) / "x.md")


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: tests/tenant.test.ts
import { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, rmSync, writeFileSync, mkdirSync } from 'node:fs';
import { tmpdir, hostname } from 'node:os';
import { join } from 'node:path';
import { tenantId, tenantPath, __resetForTests } from '../src/tenant.ts';

const localHost = hostname().split('.')[0];

// Snapshot env vars; restore on each cleanup so tests stay isolated.
let savedEnv: Record<string, string | undefined>;

function snapshotEnv(): Record<string, string | undefined> {
	return {
		SUTANDO_TENANT_ID: process.env.SUTANDO_TENANT_ID,
		SUTANDO_PRIVATE_DIR: process.env.SUTANDO_PRIVATE_DIR,
		SUTANDO_WORKSPACE: process.env.SUTANDO_WORKSPACE,
	};
}

function restoreEnv(snap: Record<string, string | undefined>): void {
	for (const k of Object.keys(snap)) {
		if (snap[k] === undefined) delete process.env[k];
		else process.env[k] = snap[k];
	}
}

describe('tenantId resolution', () => {
	let tempDir: string;

	beforeEach(() => {
		savedEnv = snapshotEnv();
		__resetForTests();
		tempDir = mkdtempSync(join(tmpdir(), 'sutando-tenant-'));
		delete process.env.SUTANDO_TENANT_ID;
		delete process.env.SUTANDO_PRIVATE_DIR;
		process.env.SUTANDO_WORKSPACE = tempDir;
	});

	afterEach(() => {
		restoreEnv(savedEnv);
		__resetForTests();
		try { rmSync(tempDir, { recursive: true, force: true }); } catch {}
	});

	it('falls back to "default" with no env, no state file', () => {
		assert.equal(tenantId(), 'default');
	});

	it('reads SUTANDO_TENANT_ID env var when set + valid', () => {
		process.env.SUTANDO_TENANT_ID = 'acme';
		assert.equal(tenantId(), 'acme');
	});

	it('rejects invalid SUTANDO_TENANT_ID + falls back', () => {
		process.env.SUTANDO_TENANT_ID = 'BAD ID!';
		assert.equal(tenantId(), 'default');
	});

	it('reads state/tenant.json when env unset', () => {
		mkdirSync(join(tempDir, 'state'), { recursive: true });
		writeFileSync(join(tempDir, 'state', 'tenant.json'), JSON.stringify({ id: 'beta-tenant' }));
		assert.equal(tenantId(), 'beta-tenant');
	});

	it('env wins over state file', () => {
		mkdirSync(join(tempDir, 'state'), { recursive: true });
		writeFileSync(join(tempDir, 'state', 'tenant.json'), JSON.stringify({ id: 'from-file' }));
		process.env.SUTANDO_TENANT_ID = 'from-env';
		assert.equal(tenantId(), 'from-env');
	});

	it('corrupt state/tenant.json → fall back to default', () => {
		mkdirSync(join(tempDir, 'state'), { recursive: true });
		writeFileSync(join(tempDir, 'state', 'tenant.json'), 'not json {{');
		assert.equal(tenantId(), 'default');
	});

	it('cached after first call', () => {
		process.env.SUTANDO_TENANT_ID = 'first';
		assert.equal(tenantId(), 'first');
		// Mutate env after caching; result unchanged until reset.
		process.env.SUTANDO_TENANT_ID = 'second';
		assert.equal(tenantId(), 'first');
		__resetForTests();
		assert.equal(tenantId(), 'second');
	});
});

describe('tenantPath resolution', () => {
	let tempPrivate: string;
	let tempWs: string;

	beforeEach(() => {
		savedEnv = snapshotEnv();
		__resetForTests();
		tempPrivate = mkdtempSync(join(tmpdir(), 'sutando-priv-'));
		tempWs = mkdtempSync(join(tmpdir(), 'sutando-ws-'));
		process.env.SUTANDO_PRIVATE_DIR = tempPrivate;
		process.env.SUTANDO_WORKSPACE = tempWs;
		delete process.env.SUTANDO_TENANT_ID;
	});

	afterEach(() => {
		restoreEnv(savedEnv);
		__resetForTests();
		try { rmSync(tempPrivate, { recursive: true, force: true }); } catch {}
		try { rmSync(tempWs, { recursive: true, force: true }); } catch {}
	});

	it('shared scope: returns tenant-<id>/<file> when private dir + tenant present', () => {
		process.env.SUTANDO_TENANT_ID = 'acme';
		const tenantDir = join(tempPrivate, 'tenant-acme');
		mkdirSync(tenantDir, { recursive: true });
		writeFileSync(join(tenantDir, 'build_log.md'), 'acme log');
		const p = tenantPath('build_log.md', 'shared');
		assert.equal(p, join(tenantDir, 'build_log.md'));
	});

	it('device scope: returns tenant-<id>/machine-<host>/<file>', () => {
		process.env.SUTANDO_TENANT_ID = 'acme';
		const tenantDir = join(tempPrivate, 'tenant-acme');
		const machineDir = join(tenantDir, 'machine-' + localHost);
		mkdirSync(machineDir, { recursive: true });
		writeFileSync(join(machineDir, 'pending-questions.md'), 'q');
		const p = tenantPath('pending-questions.md', 'device');
		assert.equal(p, join(machineDir, 'pending-questions.md'));
	});

	it('legacy fallback: default tenant + pre-migration files at top level', () => {
		// Old install: $SUTANDO_PRIVATE_DIR/notes.md (no tenant prefix).
		writeFileSync(join(tempPrivate, 'notes.md'), 'legacy notes');
		const p = tenantPath('notes.md', 'shared');
		assert.equal(p, join(tempPrivate, 'notes.md'));
	});

	it('legacy fallback for device scope: machine-<host>/<file> at top level', () => {
		const legacyMachine = join(tempPrivate, 'machine-' + localHost);
		mkdirSync(legacyMachine, { recursive: true });
		writeFileSync(join(legacyMachine, 'stand-identity.json'), '{}');
		const p = tenantPath('stand-identity.json', 'device');
		assert.equal(p, join(legacyMachine, 'stand-identity.json'));
	});

	it('workspace fallback when nothing exists in private dir', () => {
		writeFileSync(join(tempWs, 'README.md'), 'in workspace');
		const p = tenantPath('README.md', 'shared');
		assert.equal(p, join(tempWs, 'README.md'));
	});

	it('non-existent target returns preferred private path for graceful caller existsSync', () => {
		process.env.SUTANDO_TENANT_ID = 'acme';
		const p = tenantPath('does-not-exist.md', 'shared');
		assert.equal(p, join(tempPrivate, 'tenant-acme', 'does-not-exist.md'));
	});

	it('no SUTANDO_PRIVATE_DIR → workspace path always', () => {
		delete process.env.SUTANDO_PRIVATE_DIR;
		writeFileSync(join(tempWs, 'x.md'), 'x');
		assert.equal(tenantPath('x.md', 'shared'), join(tempWs, 'x.md'));
	});
});
